File: algorithms/0070.Climbing Stairs/solution.rs
pub struct Solution;

impl Solution {
    pub fn climb_stairs(n: i32) -> i32 {
        let mut pair = (0, 1);
        for _ in 0..n {
            let (prev, cur) = pair;
            pair = (cur, prev + cur);
        }
        pair.1
    }
}

#[test]
fn test1() {
    assert_eq!(Solution::climb_stairs(2), 2);
}

#[test]
fn test2() {
    assert_eq!(Solution::climb_stairs(3), 3);
}

#[test]
fn test3() {
    assert_eq!(Solution::climb_stairs(45), 1836311903);
}

File: algorithms/0070.Climbing Stairs/Solution.cs
namespace ClimbingStairs
{
    public class Solution
    {
        public int ClimbStairs(int n)
        {
            var arr = new int[n + 2];
            arr[0] = 1;
            for (var i = 0; i < n; i++)
            {
                arr[i + 1] += arr[i];
                arr[i + 2] += arr[i];
            }

            return arr[n];
        }

        public int ClimbStairs2(int n)
        {
            var (prev, next) = (0, 1);
            for (var i = 0; i < n; i++)
            {
                (prev, next) = (next, next + prev);
            }

            return next;
        }
    }
}
